_sips/sips/2017-10-25-adding-prefix-types.md

```diff
@@ -0,0 +1,163 @@
+---
+layout: sip
+discourse: true
+title: SIP-NN - Adding prefix types
+
+vote-status: pending
+permalink: /sips/:title.html
+---
+
+**By: Oron Port**
+
+## History
+
+| Date          | Version                                  |
+| ------------- | ---------------------------------------- |
+| Oct 25th 2017 | Split prefix types from [SIP33](http://docs.scala-lang.org/sips/priority-based-infix-type-precedence.html), and emphasize motivation |
+
+
+Your feedback is welcome! If you're interested in discussing this proposal, head over to [this](https://contributors.scala-lang.org/t/sip-nn-make-infix-type-alias-precedence-like-expression-operator-precedence/471) Scala Contributors thread and let me know what you think.
+
+---
+
+## Introduction
+Currently scala supports unary prefix operators (`-`, `+`, `~`, `!`) for expressions (e.g., `def unary_-`) and does not support prefix types.  See the Scala specification [Prefix Operations](http://scala-lang.org/files/archive/spec/2.12/06-expressions.html#prefix-operations) section.
+
+**Prefix expression vs. prefix type example**:
+
+```scala
+object PrefixExpression {
+  case class Nummy(expand : String) {
+    def unary_- : Nummy = Nummy(s"-$this")
+    def unary_~ : Nummy = Nummy(s"~$this")
+    def unary_! : Nummy = Nummy(s"!$this")
+    def unary_+ : Nummy = Nummy(s"+$this")
+  }
+  object N extends Nummy("N")
+  val n1 = -N
+  val n2 = ~N
+  val n3 = !N
+  val n4 = +N
+}
+object NonExistingPrefixTypes {
+  trait unary_-[A]
+  trait unary_~[A]
+  trait unary_![A]
+  trait unary_+[A]
+  trait N
+  type N1 = -N //Not working
+  type N2 = ~N //Not working
+  type N3 = !N //Not working
+  type N4 = +N //Not working
+}
+```
+
+---
+
+## Proposal
+Add support for prefix types, which is equivalent to the prefix operations for expressions.
+
+```
+PrefixType ::= [`-' | `+' | `~' | `!'] SimpleType
+CompoundType ::= PrefixType
+              |  AnnotType {with AnnotType} [Refinement]
+              |  Refinement
+```
+
+---
+
+## Motivation
+Developers expect terms and types to be expressed the same for mathematical and logical operations.
+
+### Motivating examples
+
+The [singleton-ops library](https://github.com/fthomas/singleton-ops) with [Typelevel Scala](https://github.com/typelevel/scala) (which implemented [SIP-23](http://docs.scala-lang.org/sips/pending/42.type.html)) enable developers to express literal type operations more intuitively. For example:
+
+```scala
```

Comment:
Can you add text that complements this snippet of code? For examples, saying "let's define foo and bar for..." or something similar. You can draw inspiration from SIP-35 Opaque types.

Reply:
:+1: